File: 278. First Bad Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
def isBadVersion(version):
    ##This API gived from the question to check the bad version.
    pass

class Solution:
    def firstBadVersion(self, n):

        def BS(left, right):
            if not left <= right:
                return left
            midd = (left + right) // 2
            if isBadVersion(midd):
                return BS(left, midd - 1)
            else:
                return BS(midd + 1, right)

        return BS(0, n - 1)

File: 278. First Bad Version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        long l=0, r=n;
        while (l<=r){
            long mid = (l+r)/2;
            if (isBadVersion(mid)){
                r=mid-1;
            }else{
                l=mid+1;
            }
        }
        return l;
    }
};
